File: src/main/java/net/atos/zac/app/identity/IdentityRESTService.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2023-2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.app.identity;

import java.util.List;

import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

import net.atos.zac.app.identity.converter.RESTGroupConverter;
import net.atos.zac.app.identity.converter.RESTUserConverter;
import net.atos.zac.app.identity.model.RESTGroup;
import net.atos.zac.app.identity.model.RESTLoggedInUser;
import net.atos.zac.app.identity.model.RESTUser;
import net.atos.zac.authentication.LoggedInUser;
import net.atos.zac.identity.IdentityService;
import net.atos.zac.identity.model.Group;
import net.atos.zac.identity.model.User;

@Singleton
@Path("identity")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class IdentityRESTService {

  @Inject private RESTGroupConverter groupConverter;

  @Inject private RESTUserConverter userConverter;

  @Inject private IdentityService identityService;

  @Inject private Instance<LoggedInUser> loggedInUserInstance;

  @GET
  @Path("groups")
  public List<RESTGroup> listGroups() {
    final List<Group> groups = identityService.listGroups();
    return groupConverter.convertGroups(groups);
  }

  @GET
  @Path("groups/{groupId}/users")
  public List<RESTUser> listUsersInGroup(@PathParam("groupId") final String groupId) {
    final List<User> users = identityService.listUsersInGroup(groupId);
    return userConverter.convertUsers(users);
  }

  @GET
  @Path("users")
  public List<RESTUser> listUsers() {
    final List<User> users = identityService.listUsers();
    return userConverter.convertUsers(users);
  }

  @GET
  @Path("loggedInUser")
  public RESTLoggedInUser readLoggedInUser() {
    return userConverter.convertLoggedInUser(loggedInUserInstance.get());
  }
}

File: src/main/java/net/atos/zac/app/identity/IdentityRestService.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.app.identity

import jakarta.enterprise.inject.Instance
import jakarta.inject.Inject
import jakarta.inject.Singleton
import jakarta.ws.rs.Consumes
import jakarta.ws.rs.GET
import jakarta.ws.rs.Path
import jakarta.ws.rs.PathParam
import jakarta.ws.rs.Produces
import jakarta.ws.rs.core.MediaType
import net.atos.zac.app.identity.converter.RestGroupConverter
import net.atos.zac.app.identity.converter.RestUserConverter
import net.atos.zac.app.identity.model.RestGroup
import net.atos.zac.app.identity.model.RestLoggedInUser
import net.atos.zac.app.identity.model.RestUser
import net.atos.zac.authentication.LoggedInUser
import net.atos.zac.identity.IdentityService
import nl.lifely.zac.util.AllOpen
import nl.lifely.zac.util.NoArgConstructor

@Singleton
@Path("identity")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
@AllOpen
@NoArgConstructor
class IdentityRestService @Inject constructor(
    private val groupConverter: RestGroupConverter,
    private val userConverter: RestUserConverter,
    private val identityService: IdentityService,
    private val loggedInUserInstance: Instance<LoggedInUser>
) {
    @GET
    @Path("groups")
    fun listGroups(): List<RestGroup> =
        identityService.listGroups().let {
            groupConverter.convertGroups(it)
        }

    @GET
    @Path("groups/{groupId}/users")
    fun listUsersInGroup(@PathParam("groupId") groupId: String): List<RestUser> =
        identityService.listUsersInGroup(groupId).let {
            userConverter.convertUsers(it)
        }

    @GET
    @Path("users")
    fun listUsers(): List<RestUser> =
        identityService.listUsers().let {
            return userConverter.convertUsers(it)
        }

    @GET
    @Path("loggedInUser")
    fun readLoggedInUser(): RestLoggedInUser =
        userConverter.convertLoggedInUser(loggedInUserInstance.get())
}
